executor/builder.go

```diff
@@ -3941,6 +3964,9 @@ func (builder *dataReaderBuilder) buildTableReaderForIndexJoin(ctx context.Conte
 		}
 		for _, p := range partitions {
 			pid := p.GetPhysicalID()
+			if _, ok := explicitPartitionSelection[pid]; !ok {
+				continue
+			}
```

Comment:
I think the partitions are already pruned by `partitionProcessor.convertToIntSlice()` indirectly through `partitionPruning()`.
Thus, I think the code branches where `partitionPruning()` is not called may have the problem.